Enable "Direct Intramodule Dependencies" By Default

CodaFi opened the pull request:
"Private Intransitive Dependencies" differ from the status quo by no
longer requiring the concept of a "cascading dependency edge". This is
because the request evaluator automatically tracks, records, and replays
the names looked up while a given file is being processed by the
frontend. To remove transitivity, each primary file is charged for *all* name
lookups that occur while it is being processed.  Further, because of
the replay step, lookups hidden behind cached requests are now entirely
visible to the dependency tracking code.

The net result is that all formerly implicit transitivity in the
dependency graph has been made completely explicit and direct. This
establishes a tighter overall dependency structure for each individual
file, and results in a remarkable decrease in the amount of
files that are rebuilt for any particular change.

This feature can be disabled via
-disable-experimental-private-intransitive-dependencies, which will
cause a return to the cascading status quo.

CodaFi committed:
Add Flags to Enable or Disable EnableExperientalPrivateIntransitiveDependencies

Turn the existing flag into an on-off switch and migrate all the tests in preparation for it being on by default

CodaFi committed:
Enable ExperientalPrivateIntransitiveDependencies By Default

"Private Intransitive Dependencies" differ from the status quo by no
longer requiring the concept of a "cascading dependency edge". This is
because the request evaluator automatically tracks, records, and replays
the names looked up while a given file is being processed by the
frontend. To remove transitivity from the swiftdeps files, each
primary file processed by the Swift frontend is charged for *all* name
lookups that occur while it is being processed.  Further, because of
the replay step, lookups hidden behind cached requests are now entirely
visible to the dependency tracking code.

The net result is that all formerly implicit transitivity in the
dependency graph has been made completely explicit and direct. This
establishes a tighter overall dependency structure for each individual
file, and results in a remarkable decrease in the amount of
files that are rebuilt for any particular change.

This feature can be disabled via
-disable-experimental-private-intransitive-dependencies, which will
cause a return to the cascading status quo.

CodaFi commented:
@swift-ci smoke test

CodaFi commented on test/Frontend/PrivateFingerprints/class-fingerprint.swift:
```diff
@@ -0,0 +1,73 @@
+// RUN: %FileCheck -check-prefix=CHECK-MAINAB-RECOMPILED %s < %t/output2
+
+// CHECK-MAINAB-RECOMPILED: Queuing (initial): {compile: definesAB.o <= definesAB.swift}
+// CHECK-MAINAB-RECOMPILED: Queuing because of dependencies discovered later: {compile: usesA.o <= usesA.swift}
```
Just to call this out - since the other `PrivateFingerprints` tests exhibit a similar behavior change - this demonstrates a clear win over the status quo. Each of these tests changes the nominal type `A`, but with private dependencies, we correctly decline to queue the users of the unrelated type `B`.

davidungar replied:
Yes, excellent!

davidungar reviewed (approved):
Great!

davidungar commented on lib/Frontend/CompilerInvocation.cpp:
```diff
@@ -435,8 +435,10 @@ static bool ParseLangArgs(LangOptions &Opts, ArgList &Args,
+  Opts.EnableExperientalPrivateIntransitiveDependencies =
+      Args.hasFlag(OPT_enable_experimental_private_intransitive_dependencies,
+                   OPT_disable_experimental_private_intransitive_dependencies,
+                   /*default*/ true);
```
Suggestion: if instead of "true" you write "Opts.EnableExperimental..." here, then the default is only defined in one place, the .h file.

hamishknight commented:
Should we drop "experimental" from the name if we're enabling it by default?

CodaFi committed:
[NFC] Rename Flags

-enable-experimental-private-intransitive-dependencies -> -enable-direct-intramodule-dependencies
-disable-experimental-private-intransitive-dependencies -> -disable-direct-intramodule-dependencies

While we're here, rename DependencyCollector::Mode's constants and clean
up the documentation.

CodaFi commented:
@swift-ci smoke test

CodaFi commented:
@swift-ci smoke test

CodaFi commented:
@swift-ci smoke test

CodaFi commented:
@swift-ci smoke test

CodaFi commented:
@swift-ci clean smoke test

CodaFi commented:
@swift-ci clean smoke test

CodaFi commented:
@swift-ci smoke test macOS platform

CodaFi commented:
@swift-ci smoke test macOS platform

CodaFi commented:
@swift-ci smoke test Linux platform

CodaFi commented:
@swift-ci smoke test Linux platform

CodaFi commented:
@swift-ci smoke test macOS
